Color filenames based on file type

jimhester opened the pull request:
So this addresses #14, but has at least a couple of drawbacks.

1. Colors are only output if tibble is loaded (could also be a positive depending on your outlook on things)
2. Because dplyr verbs to do dispatch on `[` methods, the attribute is not subset appropriately when using dplyr verbs and the colors are incorrect.

I think this feature is nice, but not sure if it is worth pursuing further if it is not going to work in all cases.

hadley commented:
Can you give me a screenshot of `dir_info()`?

jimhester commented:
One alternative implementation that I think would work with the dplyr verbs and purrr is to store the paths as a list column. I did not do this because of storage efficiency, but I guess it might be the only way forward.

hadley commented:
Yeah, it might be better to leave this until dplyr/vctrs does better at preserving classes.

jimhester commented:
https://github.com/r-lib/fs/pull/23/commits/91e64d27a1d8b31dd8e6aae4e4635bcf56fd3d8d uses list columns, which do work with dplyr. I think it is worth going this route for now and we can maybe switch the implementation in the future once there is better support.

hadley commented:
That'll only work with `dir_info()`, right? So we lose the nice colouring for `dir_list()`, but that's probably a small price to pay

jimhester commented:
We don't do a `stat` for `dir_list()`, so we don't have all the information we would need to color the names. We could potentially do the stat on demand in that case, or leave `dir_list()` as a simple character vector as it is currently.

Here is how it looks in RStudio with the material theme and the LS_COLORS envirionment variable set with the file from https://github.com/trapd00r/LS_COLORS/blob/master/LS_COLORS

![screenshot 2017-12-28 11 53 52](https://user-images.githubusercontent.com/205275/34417306-ca85e458-ebc5-11e7-8db8-ac31bfa3c1a6.png)

hadley commented:
Oooh, that's pretty nice.  

I wonder if we should provide standard colours for "R" extensions?

jimhester commented:
I made a major change in how this works. Now we do not store any mode information with the path, we simply stat the paths on-demand when we are colorizing paths. This means we stat files twice for `file_info()` / `dir_info()`, but it also means we can use this coloring everywhere.

I also added a `path_tidy()` function, which outputs tidy paths and changed all the functions in **fs** to return them.
> Tidy paths always expand `~`, use `/` to delimit directories, never have multiple `/` or
trailing `/` and have colourised output based on the file type.

It is true this will result in more stat calls in general but they are cheap, and will only happen when printing to the console.

hadley commented on DESCRIPTION:
```diff
@@ -26,4 +26,8 @@ URL: https://github.com/r-lib/fs
+    covr,
+    pillar,
+    crayon
+Remotes:
```
pillar is on CRAN, fyi

hadley commented on R/path.R:
```diff
@@ -33,13 +32,40 @@ path_norm <- function(path) {
+#' Tidy paths
+#'
+#' untidy paths are all different, tidy paths are all the same.
+#'
```
I think you can drop the new line here

hadley commented on R/path.R:
```diff
@@ -33,13 +32,40 @@ path_norm <- function(path) {
+  # Remove trailing / from paths
+  path <- sub("/$", "", path)
+
+  path
```
I think it would be clearer to do `new_fs_filename` here?

hadley commented on R/path.R:
```diff
@@ -33,13 +32,40 @@ path_norm <- function(path) {
 
 #' Provide the path to the users home directory
 #' @export
 path_home <- function() {
```
Not really germane to this PR, but I think we should pass `...` onto to `path()`

hadley commented on R/utils.R:
```diff
@@ -20,9 +20,15 @@ auto_name <- function(names) {
+# tests.
+compare.fs_filename <- function(x, y) {
+  if (identical(class(y), "character")) {
+    class(x) <- "character"
```
Setting the class of a base type is a slightly odd thing to do.

jimhester replied:
Yeah you are right, I can just use `class(x) <- NULL` here and get the same result.

hadley commented:
Will also close #20 and maybe #21.

hadley commented:
And #2
